File: gamechannel/signatures.cpp
// Copyright (C) 2019-2022 The Xaya developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include "signatures.hpp"

#include <xayautil/base64.hpp>
#include <xayautil/hash.hpp>

#include <glog/logging.h>

#include <sstream>

namespace xaya
{

std::string
GetChannelSignatureMessage (const std::string& gameId,
                            const uint256& channelId,
                            const proto::ChannelMetadata& meta,
                            const std::string& topic,
                            const std::string& data)
{
  for (const char t : topic)
    CHECK ((t >= '0' && t <= '9')
              || (t >= 'A' && t <= 'Z')
              || (t >= 'a' && t <= 'z'))
      << "Topic string contains invalid character: " << topic;

  std::ostringstream res;
  res << "Game-Channel Signature\n"
      << "Game ID: " << gameId << "\n"
      << "Channel: " << channelId.ToHex () << "\n"
      << "Reinit: " << EncodeBase64 (meta.reinit ()) << "\n"
      << "Topic: " << topic << "\n"
      << "Data Hash: " << SHA256::Hash (data).ToHex ();

  return res.str ();
}

std::set<int>
VerifyParticipantSignatures (const SignatureVerifier& verifier,
                             const std::string& gameId,
                             const uint256& channelId,
                             const proto::ChannelMetadata& meta,
                             const std::string& topic,
                             const proto::SignedData& data)
{
  const auto msg
      = GetChannelSignatureMessage (gameId, channelId, meta,
                                    topic, data.data ());

  std::set<std::string> addresses;
  for (const auto& sgn : data.signatures ())
    addresses.emplace (verifier.RecoverSigner (msg, sgn));

  std::set<int> res;
  for (int i = 0; i < meta.participants_size (); ++i)
    if (addresses.count (meta.participants (i).address ()) > 0)
      res.emplace (i);

  return res;
}

bool
SignDataForParticipant (SignatureSigner& signer,
                        const std::string& gameId,
                        const uint256& channelId,
                        const proto::ChannelMetadata& meta,
                        const std::string& topic,
                        const int index,
                        proto::SignedData& data)
{
  CHECK_GE (index, 0);
  CHECK_LT (index, meta.participants_size ());
  const std::string& addr = meta.participants (index).address ();
  LOG (INFO) << "Trying to sign data with address " << addr << "...";

  if (addr != signer.GetAddress ())
    {
      LOG (ERROR) << "The provided signer is for a different address";
      return false;
    }

  const auto msg
      = GetChannelSignatureMessage (gameId, channelId, meta,
                                    topic, data.data ());
  data.add_signatures (signer.SignMessage (msg));
  return true;
}

} // namespace xaya

File: gamechannel/signatures.py
# Copyright (C) 2019-2022 The Xaya developers
# Distributed under the MIT software license, see the accompanying
# file COPYING or http://www.opensource.org/licenses/mit-license.php.

"""
Utility methods for working with SignedData objects from game channels.
"""

from .proto import (metadata_pb2, signatures_pb2)

import base64
import codecs
import hashlib


def getChannelMessage (gameId, channelId, meta, topic, data):
  """
  Returns the raw message that is signed (as string, using signmessage)
  for signing the given data in a channel with the given ID and topic.

  The channelId is given as 32 raw bytes that encode the uint256 directly
  (in big endian, i.e. the same way it is represented in hex).
  """

  assert len (channelId) == 32

  reinit = codecs.decode (base64.b64encode (meta.reinit), "ascii")

  return "Game-Channel Signature\n" \
          + ("Game ID: %s\n" % gameId) \
          + ("Channel: %s\n" % channelId.hex ()) \
          + ("Reinit: %s\n" % reinit) \
          + ("Topic: %s\n" % topic) \
          + ("Data Hash: %s" % hashlib.sha256 (data).hexdigest ())


def createForChannel (rpc, gameId, channel, topic, data):
  """
  Constructs a SignedData protocol buffer instance that contains the given
  data and signatures for all participants of the channel for which the private
  key is known through the RPC connection.

  channel should be an object similar to how game channels are represented
  in the game state JSON.  In particular, it should have an "id" field with
  the channel ID as hex, and meta.proto should be the base64-encoded serialised
  metadata proto.
  """

  res = signatures_pb2.SignedData ()
  res.data = data

  channelId = codecs.decode (channel["id"], "hex")
  meta = metadata_pb2.ChannelMetadata ()
  meta.ParseFromString (base64.b64decode (channel["meta"]["proto"]))
  msg = getChannelMessage (gameId, channelId, meta, topic, data)

  for p in meta.participants:
    valid = rpc.validateaddress (p.address)
    if not valid["isvalid"]:
      continue
    info = rpc.getaddressinfo (p.address)
    if info["ismine"]:
      sgn = rpc.signmessage (p.address, msg)
      res.signatures.append (sgn)

  return res
